streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java

```diff
@@ -0,0 +1,139 @@
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import java.util.Collection;
+import java.util.List;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.processor.internals.StreamThread.State;
+import org.apache.kafka.streams.processor.internals.assignment.AssignorError;
+import org.slf4j.Logger;
+
+public class StreamsRebalanceListener implements ConsumerRebalanceListener {
+
+    private final Time time;
+    private final TaskManager taskManager;
+    private final StreamThread streamThread;
+    private final Logger log;
+
+    StreamsRebalanceListener(final Time time,
+        final TaskManager taskManager,
+        final StreamThread streamThread,
+        final Logger log) {
+        this.time = time;
+        this.taskManager = taskManager;
+        this.streamThread = streamThread;
+        this.log = log;
+    }
+
+    @Override
+    public void onPartitionsAssigned(final Collection<TopicPartition> assignedPartitions) {
+        log.debug("at state {}: partitions {} assigned at the end of consumer rebalance.\n" +
+                "\tpreviously assigned active tasks: {}\n" +
+                "\tpreviously assigned standby tasks: {}\n",
+            streamThread.state(),
+            assignedPartitions,
+            taskManager.activeTaskIds(),
+            taskManager.standbyTaskIds());
+
+        if (streamThread.getAssignmentErrorCode() == AssignorError.INCOMPLETE_SOURCE_TOPIC_METADATA.code()) {
+            log.error("Received error code {} - shutdown", streamThread.getAssignmentErrorCode());
+            streamThread.shutdown();
+            return;
+        }
+        final long start = time.milliseconds();
+        List<TopicPartition> revokedChangelogPartitions = null;
+        try {
+            if (streamThread.setState(State.PARTITIONS_ASSIGNED) == null) {
+                log.debug(
+                    "Skipping task creation in rebalance because we are already in {} state.",
+                    streamThread.state()
+                );
+            } else if (streamThread.getAssignmentErrorCode() != AssignorError.NONE.code()) {
+                log.debug(
+                    "Encountered assignment error during partition assignment: {}. Skipping task initialization",
+                    streamThread.getAssignmentErrorCode()
+                );
+            } else {
+                // Close reassigned tasks before initializing new ones as we may have suspended active
+                // tasks that become standbys or vice versa
+                log.debug("Closing any no-longer assigned tasks.");
+                revokedChangelogPartitions = taskManager.closeRevokedTasks();
+                log.debug("Creating new tasks based on assignment.");
+                taskManager.createTasks(assignedPartitions);
```

Comment:
See my other comment: I feel it's cleaner to execute `closeSuspendedActiveTasks` that and `closeStandbyTasks` differently.

Reply:
Done